src/main.rs

```diff
@@ -51,25 +55,24 @@ async fn main() -> anyhow::Result<()> {
         command,
     } = Args::parse();
 
+    // Expand vector as more CRDs are created.
+    let crds = vec![kubit::resources::AppInstance::crd()];
     match &command {
-        Some(Commands::Manifests { crd_dir }) => match crd_dir {
-            Some(crd_dir) => {
-                // Expand vector as more CRDs are created.
-                for crd in vec![&kubit::resources::AppInstance::crd()] {
-                    let crd_file = format!("{}_{}.yaml", crd.spec.group, crd.spec.names.plural);
-                    let crd_path = PathBuf::from(crd_dir).join(crd_file);
-
-                    let file =
-                        File::create(crd_path).expect("Could not open AppInstances CRD file");
-
-                    serde_yaml::to_writer(&file, &kubit::resources::AppInstance::crd())?;
-                }
+        Some(Commands::Manifests { crd_dir }) => {
+            for crd in crds {
+                let out_writer = match crd_dir {
+                    Some(dir) => {
+                        let crd_file = format!("{}_{}.yaml", crd.spec.group, crd.spec.names.plural);
+                        let crd_path = PathBuf::from(dir).join(crd_file);
+                        let file =
+                            File::create(crd_path).expect("Could not open AppInstances CRD file");
+                        Box::new(file) as Box<dyn Write>
+                    }
+                    None => Box::new(stdout()) as Box<dyn Write>,
+                };
+                serde_yaml::to_writer(out_writer, &crd).unwrap();
```

Comment:
@mkmik This seems to satisfy what we want, much appreciated on the control flow inversion tip.

To me this seems quite odd though because we run the `match` on every CRD we have and set what the `out_writer` should be per iteration, even when we _know_ that as soon as it is set we are writing to a `File`, compared to `stdout` without it.

That's where my confusion was initially as it seems odd to continually check this, but I suppose the avoidance of the continually nested `match, Some ...` is better here?

---

As for the `Box` usage, I believe this is correct. Coming from [The Book](https://doc.rust-lang.org/book/ch15-01-box.html)

> When you want to own a value and you care only that it’s a type that implements a particular trait rather than being of a specific type

I.e here we only care that both `stdout` and the `File` implement the `std::io::Write` trait so we can treat them the same way.

Reply:
Awesome! Your first battle with boxed objects, the things most other languages are made of :-)

> [..] even when we know that as soon as it is set we are writing to a File, compared to stdout without it.

yes, but every iteration would have its own file, so it's only "wasteful" when it's writing to stdout.

In this case there is no point optimizing the hell out of it, but wondering how one would do it is a natural phase in the learning curve of a new language so I understand the instinct. 

If you want to avoid doing the "wasteful" operation and do the check only once, you can do something like this:


(pseudo code, doesn't compile):

```rust
        Some(Commands::Manifests { crd_dir }) => {
            let mut out_writer: Box<dyn Write> = Box::new(stdout());
            for crd in crds {
                if let Some(dir) = crd_dir {
                    let crd_file = format!("{}_{}.yaml", crd.spec.group, crd.spec.names.plural);
                    let file = File::create(dir.join(crd_file))
                        .expect("Could not open AppInstances CRD file");
                    out_writer = Box::new(file);
                };
                // The YAML delimiter is added in the event we have multiple documents.
                let out_writer = &mut out_writer;
                writeln!(out_writer, "---").unwrap();
                serde_yaml::to_writer(out_writer, &crd).unwrap();
            }
        }```

This allocates a new boxed to hold the file on every iteration, but it doesn't do any "wasteful" cycles to allocate a box to hold stdout when stdout is going to always be the same stdout throughout the iteration.

However, by doing this we unnecessarily complicated our code, which now requires us to understand the delicate rules about mutability and borrowing rules. The code above looks simple once you've written it, but the smallest mistake (like the lack of `&mut out_writer`) will make you fight errors like:

```
68 |             let mut out_writer: Box<dyn Write> = Box::new(stdout());
   |                 -------------- move occurs because `out_writer` has type `Box<dyn std::io::Write>`, which does not implement the `Copy` trait
...
74 |                     out_writer = Box::new(file);
   |                     ---------- this reinitialization might get skipped
...
77 |                 writeln!(out_writer, "---").unwrap();
   |                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^ value borrowed here after move
78 |                 serde_yaml::to_writer(out_writer, &crd).unwrap();
   |                                       ---------- value moved here, in previous iteration of loop
```

this is what makes rust harder than the average language. It's ability to make safe but high performant code.

But you don't need that all the time. This conversely means that if you don't need to squeeze the maximum out of the machine all the time you can make rust be easier for you most of the times!